Add command for fetching the latest release info from a github repo

kaffein opened the pull request:
Hey guys,

Sorry that it took so long for me to come back for this PR but I wanted to get a little bit acquainted with the project (real beginner here ^^). In the end though, it feels good to learn some new stuff ;-) and fortunately, here is the first PR I can come up with.

This one is related to issue #729 : we wanted a command for fetching infos about the latest release on a github repository.
The syntax for the command is quite simple : 

`!gh releases org-name/repo`

Here are some screenshots of how it renders :

1. When simply asking the bot to fetch the latest release info
![gh_simple_call](https://user-images.githubusercontent.com/847920/47597118-27bac300-d98c-11e8-9c11-c3f77bf1b5f3.png)

2. When asking for release infos which are, unfortunately, unavailable
![gh_releases_no_info_found](https://user-images.githubusercontent.com/847920/47597132-44ef9180-d98c-11e8-86c7-18325a663f44.png)

3. When used with the already existing `cron` command
![gh_releases_result_on_cron](https://user-images.githubusercontent.com/847920/47597143-66e91400-d98c-11e8-8e79-a0967acd5289.png)

I hope my code is not too ugly ^^ but anyway, feel free to let me know of any changes that you want me to incorporate into the PR. I'd be glad to have your feedback especially on the _Clojure idioms_ or anything else that can be improved.

thanks a lot for the opportunity o/

devth commented:
Thanks for the PR! The code looks great.

If you want, we could expand the capabilities a bit to include:

- list all known releases
- show details for a specific release

So something like:

```bash
gh releases <org>/<repo-name> # list releases for a GitHub repo
gh releases show # show info for the latest release 
gh releases show <tag> # show info for a specific release
```

If you want to work on those feel free to continue on this PR, or if not I can merge this as is.

kaffein commented:
Thanks for your feedback,

I will work on these extensions from this same PR and will poke you again for reviews when it's ready ;-)

kaffein commented:
hey guys,

I have refactored the code according to @devth suggestions

so here are some screenshots of how it renders at the moment : 

1. When listing all releases from a Github repo
![gh_repo_releases](https://user-images.githubusercontent.com/847920/47623293-0e428400-db10-11e8-993a-c1a386525746.png)

2. When fetching info for the latest release from a Github repo
![gh_latest_release](https://user-images.githubusercontent.com/847920/47623320-4944b780-db10-11e8-9caa-b77a0b3ab396.png)

3. When fetching release info from a _specific tag_
![gh_existing_release_by_tag](https://user-images.githubusercontent.com/847920/47623340-ad677b80-db10-11e8-977f-182b72a29dd3.png)

4. When fetching release info from a _specific tag_ but the tag does not exist
![gh_no_matching_release_tag](https://user-images.githubusercontent.com/847920/47623352-c6702c80-db10-11e8-899c-2864ecac8c7c.png)

There may be some ways to reduce boilerplate for these commands since displaying release infos are always almost the same. Only the _wording_ and the _information_ to be displayed change, depending on the command.

As usual, feel free to let me know if you guys think there are some other things that need to be improved or that I may have missed. 

again thanks ...

devth commented:
@kaffein awesome! I'll review tomorrow.

devth commented on src/yetibot/commands/github.clj:
```diff
@@ -136,6 +137,50 @@
+  [{[_ org-name repo] :match}]
+  (let [latest-release (gh/latest-releases org-name repo)
+        status (:status latest-release)]
+    (if (nil? status)
```
Let's use a helper function to reduce the similar code between `show-latest-release-info-cmd` and `show-release-info-by-tag-cmd`. Maybe call it `format-release`.

kaffein replied:
I'm on it ^^

kaffein commented:
Hey guys,

based on your reviews, I have done some refactoring to remove similar code.
I have also fixed a bug that I have introduced, related to how the `published_at` date was being parsed.

can you guys PTAL

thanks in advance

devth commented on src/yetibot/commands/github.clj:
```diff
@@ -137,36 +137,32 @@
-                (f/unparse date-hour-formatter published-at) author))
-      (format "No release version info found for %s/%s" org-name repo))))
+  (if-let [release (gh/latest-releases org-name repo)]
+    (format-release release org-name repo)))
```
Looks good but do we need to handle the case where `release` is nil and tell the user that we couldn't find the release? I see you already have that error handling in `format-release`, but it'll never get called of the release is nil right?

kaffein replied:
Actually, `gh/latest-releases` and `gh/release-by-tag` always return a _json result_ when called, even though there are no releases associated with a repository. So `release` always have a value.
In `format-release`, I then check for the presence of the `:status` field in the json response, to check whether information about the release has been found or not.

```
(gh/latest-releases "kaffein" "clojure-applied")
=>
{:request-time 1378,
 ...
 ...
 :status 404,
 :length -1,
 ...,
 :trace-redirects []}
```

```
(gh/release-by-tag "kaffein" "clojure-applied" "1.0.0-RC1")
=>
{:request-time 997,
 ...
 :status 404,
 ...,
 :trace-redirects []}
```

_Note_ : I probably may have used `let` here instead of `if-let`, since I can always expect to have a result from the call.

On the other hand though, the call to `gh/releases` may return `nil`. I should have used `if-let` here instead.

```
(gh/releases "kaffein" "clojure-applied")
=> nil
```

devth replied:
Yep I'd switch that to `let` if you always expect it to be non-nil.

devth commented on src/yetibot/commands/github.clj:
```diff
@@ -137,36 +137,32 @@
+          published-at (-> "YYYY-MM-dd'T'HH:mm:ssZ"
+                           (f/formatter)
+                           (f/parse (:published_at release)))]
+      (format "Release version, tagged: %s from %s/%s, was published on %s by %s" tag org-name repo
```
For `show` commands I'd like to see more data for that release. Can we render the body or description of the release? For example, this release page contains a ton of interesting info: https://github.com/helm/helm/releases/tag/v2.11.0

kaffein replied:
That makes sense, I'm on it ^^

kaffein replied:
so, this is how it renders at the moment : 

![show_cmd_with_changelogs](https://user-images.githubusercontent.com/847920/47920170-b27e4f00-deb1-11e8-810d-bd030742fab2.png)

is it cool for you guys ?
thanks

devth replied:
Yep, looks great!

kaffein commented:
thanks @devth :)
